api/envoy/config/wasm/v2/wasm.proto

```diff
@@ -0,0 +1,68 @@
+syntax = "proto3";
+
+package envoy.config.wasm.v2;
+
+option java_outer_classname = "WasmProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.config.wasm.v2";
+
+import "envoy/api/v2/core/base.proto";
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Wasm service]
+
+// Configuration for a Wasm VM.
+// [#next-free-field: 6]
+message VmConfig {
+  // An ID which will be used along with a hash of the wasm code (or null_vm_id) to determine which
+  // VM will be used for the plugin. All plugins which use the same vm_id and code will use the same
+  // VM. May be left blank.
+  string vm_id = 1;
+
+  // The Wasm runtime type (see source/extensions/common/wasm/well_known_names.h).
+  string runtime = 2;
+
+  // The Wasm code that Envoy will execute.
+  api.v2.core.AsyncDataSource code = 3;
+
+  // The Wasm configuration string used on initialization of a new VM (proxy_onStart).
+  string configuration = 4;
+
+  // Allow the wasm file to include pre-compiled code on VMs which support it.
+  bool allow_precompiled = 5;
+}
+
+// Base Configuration for Wasm Plugins e.g. filters and services.
+message PluginConfig {
```

Comment:
name this `ExtensionConfig`? if this is going to be filters / loggers, "extension" for consistency.

Reply:
I have found that naming is often contentions and orthogonal to the overall design.  Let's create an issue for naming this and then we can discuss it separately from the design.  In this case the word "extension" in Envoy already means something very specific (source/extension) and overloading it here with a different meaning is likely to be confusing.

Reply:
I think this PR is where we are doing this discussion :) If not now, when? :)

Reply:
OK.  I prefer plugin over extension for the reasons above, @htuch @mattklein123 ?

Reply:
I prefer "WASM extension" or "WASM sub-extension". The rationale is that plugin is a synonym for extension, and it becomes conceptually complicated to have many similar words (which sometimes get used even more loosely in everyday English). Having a single term, "extension", and then qualifying it with "WASM" or "sub-", is more precise IMHO.

Reply:
(FWIW we already have a lot of confusion over "filter" vs "extension", it pays to be really careful when establishing this terminology, which is a 20/20 hindsight kind of observation)

Reply:
There is potential confusion between "WASM extension" (lower case, your proposal) vs The WASM Extension" (upper case) in that the WASM support code is in source/extensions and in particular the WASM service code is in source/extensions/wasm.  The net is that if we got that direction then the word "plugin" must be substituted with "a WASM Extension" and there is no good way to use the definite article "the" without being ambiguous.  Which, as someone who is going to be writing a lot of docs with this term, is problematic.

"Wasm extension module"?  Or just "Wasm module" ?

Reply:
"Wasm module" works for me. We can sed "Plugin" to "WasmModule" and call it a day I think, unless others want to bike shed more.

Reply:
WasmModule has one wart: it technically has a 1:1 relationship with a VM.  There are proposals to change that, but they have not been accepted yet.  Technically that means that to WASM folks, a WasmModule would not correspond to one of a set of plugins which (currently) must be linked into a single "Wasm Module" for loading into a VM, but the fully linked object.  

Unfortunately that is not what we are referring to here since this is specifically the wasm code associated with a particular extensions point which may for efficiency purposes be linked into the same "Wasm Module" to be loaded into a VM to be used for a set of extension points.

I am not sure how much confusing this will cause.  It has already caused some confusion during our discussions about the istio plugins.

Reply:
Agreed, thanks for the explanation. Let's iterate around this in the doc and we can circle back here with the agreed naming.

Reply:
As per discussions on the doc, we don't have a better name at this time.  We can do a rename if/when we do come up with one.